src/dotenv/main.py

```diff
@@ -299,17 +299,18 @@ def _is_interactive():
 
 
 def load_dotenv(dotenv_path=None, stream=None, verbose=False, override=False, interpolate=True, **kwargs):
-    # type: (Union[Text, _PathLike, None], Optional[_StringIO], bool, bool, bool, Union[None, Text]) -> bool
+    # type: (Optional[Union[Text, _PathLike, _StringIO]], Optional[_StringIO], bool, bool, bool, Optional[Text]) -> bool
     """Parse a .env file and then load all the variables found as environment variables.
 
     - *dotenv_path*: absolute or relative path to .env file.
     - *stream*: `StringIO` object with .env content.
-    - *verbose*: whether to output the warnings related to missing .env file etc. Defaults to `False`.
+    - *verbose*: whether to output log messages related to loading .env files etc. Defaults to `False`.
     - *override*: where to override the system environment variables with the variables in `.env` file.
                   Defaults to `False`.
     """
-    f = dotenv_path or stream or find_dotenv()
-    return DotEnv(f, verbose=verbose, interpolate=interpolate, **kwargs).set_as_environment_variables(override=override)
+    f = find_dotenv(filename=str(dotenv_path)) or stream or find_dotenv()
```

Comment:
Yes. Wouldn't that be the intended behavior? Do you envision a use case in which a user would run `load_dotenv(".env")`, but they wouldn't want to actually find the .env file?

Reply:
I think `find_dotenv(filename=".env")` stops walking up the directories when it finds the file, right? Given a directory structure like this (adapted from the `prepare_file_hierarchy` test in test_main.py):

```text

project_root
└── src
    ├── package1
    │   ├── package1_child1
    │   │   └── package1_child2
    │   └── .env
    └── package2
        ├── package2_child1
        │   └── package2_child2
        └── .env
```

If `find_dotenv(".env")` is run from one of the child directories in `package2`, it will stop when it finds `package2/.env`. Is this correct?

Reply:
The other advantage to using `find_dotenv()` here is that it becomes easier to manage the return value of `load_dotenv`. If the user runs `load_dotenv(".does_not_exist")`, `f = dotenv_path` will evaluate to `f = ".does_not_exist"`, but `f = find_dotenv(".does_not_exist")` will evaluate to `f = ""`, same as the third option in the conditional expression (`find_dotenv()`).

Reply:
This is correct but the use case I had in mind was the following, where `load_dotenv(".env")` is called in `sub.py` but the `.env` is absent in `sub_project` (e.g. because the developer didn't use one yet).

```
project
├── .env
├── manage.py
└── sub_project
    └── sub.py
```

Perhaps it is not very common but it is still a breaking change if `find_dotenv` is applied by default by `load_dotenv`.

We've introduced breaking changes many times in this project, but each time we did it, I believe the trade-off was in favor of the change.  In this case, however, I'm not convinced it's worth changing the behavior, at least for now.  I don't think saving 2 or 3 lines of code is enough to justify it.

Reply:
> the use case I had in mind was the following, where `load_dotenv(".env")` is called in `sub.py` but the `.env` is absent in `sub_project`

If the user:

- Has a file at `project/.env`, but doesn't want to load it
- Doesn't have a file at `sub_project/.env`, but wants to load it
- Runs `load_dotenv(".env")` in `sub.py`, expecting to load the non-existent file

That sounds like a situation in which they should be notified of the problem, either with an exception or logger warning. Instead of providing an exception or logger warning, `load_dotenv` simply returns `True`. I find this behavior misleading, which was the reason I opened the PR originally.

> In this case, however, I'm not convinced it's worth changing the behavior, at least for now. I don't think saving 2 or 3 lines of code is enough to justify it.

The goal of this PR was to change the return behavior of `load_dotenv`. If you're not willing to consider a change, there's nothing further to discuss.

Reply:
I think there is a misunderstanding.   Changing the return behavior of `load_dotenv` might be a breaking change but it's probably worth it because the current behavior makes no sense, as some of us have pointed out in the past. Let's call it change 1.

The breaking change I was referring to when I said that it didn't seem worth it, is the fact that `load_dotenv` would now go up the hierarchy to find files when provided with a non-None `dotenv_path` argument.  Let's call it change 2. The current behavior is to not go up the file hierarchy: if a file isn't found where it's supposed to be, stop looking.

Change 1 looks really good to me, but I'm not sure about change 2.  We could discuss the value of change 2 further but my impression is that we don't need it for change 1.